File: client/src/lib/exchangeRates.ts
export interface ExchangeRateQuote {
  rate: number
  date: string
  source: 'frankfurter' | 'exchangerate-api'
}

const FRANKFURTER_API = 'https://api.frankfurter.dev/v1'
const ER_API = 'https://open.er-api.com/v6/latest/USD'

export async function fetchExchangeRate(from: string, to: string): Promise<ExchangeRateQuote> {
  if (from === to) {
    return { rate: 1, date: new Date().toISOString().slice(0, 10), source: 'frankfurter' }
  }

  try {
    return await fetchFromFrankfurter(from, to)
  } catch {
    return fetchFromExchangeRateApi(from, to)
  }
}

async function fetchFromFrankfurter(from: string, to: string): Promise<ExchangeRateQuote> {
  const response = await fetch(
    `${FRANKFURTER_API}/latest?from=${encodeURIComponent(from)}&to=${encodeURIComponent(to)}`,
  )
  if (!response.ok) {
    throw new Error('Frankfurter request failed')
  }

  const data = (await response.json()) as { date: string; rates: Record<string, number> }
  const rate = data.rates[to]
  if (rate == null) {
    throw new Error(`No Frankfurter rate for ${from} → ${to}`)
  }

  return { rate, date: data.date, source: 'frankfurter' }
}

async function fetchFromExchangeRateApi(from: string, to: string): Promise<ExchangeRateQuote> {
  const response = await fetch(ER_API)
  if (!response.ok) {
    throw new Error('Exchange rate fallback request failed')
  }

  const data = (await response.json()) as {
    result: string
    time_last_update_utc: string
    rates: Record<string, number>
  }

  if (data.result !== 'success') {
    throw new Error('Exchange rate fallback returned an error')
  }

  const fromRate = data.rates[from]
  const toRate = data.rates[to]
  if (fromRate == null || toRate == null) {
    throw new Error(`No fallback rate for ${from} → ${to}`)
  }

  return {
    rate: toRate / fromRate,
    date: data.time_last_update_utc.slice(0, 10),
    source: 'exchangerate-api',
  }
}

export function convertAmount(amount: number, rate: number): number {
  return amount * rate
}

export function rateSourceLabel(source: ExchangeRateQuote['source']): string {
  return source === 'frankfurter' ? 'ECB' : 'ExchangeRate-API'
}

File: server/app/integrations/exchange_rates.py
import time
from decimal import Decimal, ROUND_HALF_UP

import httpx

FRANKFURTER_API = "https://api.frankfurter.dev"
ER_API = "https://open.er-api.com/v6/latest/USD"
CACHE_TTL_SECONDS = 3600

_rate_cache: dict[tuple[str, str], tuple[Decimal, float]] = {}


def _quantize(amount: Decimal) -> Decimal:
    return amount.quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)


def _fetch_frankfurter(from_ccy: str, to_ccy: str) -> Decimal:
    url = f"{FRANKFURTER_API}/v2/rate/{from_ccy}/{to_ccy}"
    response = httpx.get(url, timeout=10.0)
    response.raise_for_status()
    data = response.json()
    rate = data.get("rate")
    if rate is None:
        raise ValueError(f"No Frankfurter rate for {from_ccy} → {to_ccy}")
    return Decimal(str(rate))


def _fetch_er_api(from_ccy: str, to_ccy: str) -> Decimal:
    response = httpx.get(ER_API, timeout=10.0)
    response.raise_for_status()
    data = response.json()
    if data.get("result") != "success":
        raise ValueError("Exchange rate fallback returned an error")
    rates = data.get("rates") or {}
    from_rate = rates.get(from_ccy)
    to_rate = rates.get(to_ccy)
    if from_rate is None or to_rate is None:
        raise ValueError(f"No fallback rate for {from_ccy} → {to_ccy}")
    return Decimal(str(to_rate)) / Decimal(str(from_rate))


def get_exchange_rate(from_ccy: str, to_ccy: str) -> Decimal:
    """Return multiplier to convert `from_ccy` amount into `to_ccy`."""
    from_ccy = from_ccy.upper()
    to_ccy = to_ccy.upper()
    if from_ccy == to_ccy:
        return Decimal("1")

    cache_key = (from_ccy, to_ccy)
    now = time.time()
    cached = _rate_cache.get(cache_key)
    if cached and now < cached[1]:
        return cached[0]

    try:
        rate = _fetch_frankfurter(from_ccy, to_ccy)
    except Exception:
        rate = _fetch_er_api(from_ccy, to_ccy)

    _rate_cache[cache_key] = (rate, now + CACHE_TTL_SECONDS)
    return rate


def convert_amount(amount: Decimal, from_ccy: str, to_ccy: str) -> Decimal:
    rate = get_exchange_rate(from_ccy, to_ccy)
    return _quantize(amount * rate)
